0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,353 @@
+- **`scale`** number
+  - Number of orders of magnitude smaller this unit is as compared to the unit of exchange for the asset.
+  - All units MUST be at least as precise as the unit of exchange. For example, if the asset is U.S. dollars, quantities cannot be denominated in $100 bills, since that's less precise than quantities denominated in multiples of $1. Thus, the scale MUST NOT be a negative integer.
+
+##### Example
+
+To represent $2.54 in units of cents, where the amount is multiple of $0.01:
+
+```json
+{
+  "amount": "254",
+  "scale": 2
+}
+```
+
+#### Choosing scales
+
+The unit or scale used for quantities by the settlement engine is RECOMMENDED to be the smallest denomination of the asset on the settlement system.
+
+The accounting system and settlement engine are RECOMMENDED to use the same unit or scale.
+
+#### Scale conversions
+
+Since the systems MAY use different scales, when interfacing with one another, care must be taken to ensure the system with additional precision tracks the amount leftover by the other system so it may accumulate and eventually be correctly accounted for.
+
+If one of the systems receives a request with a **[`Quantity`](#quantity-type)** denominated in a unit more precise than its unit, it SHOULD convert the quantity into its native unit. If so, the resulting amount MUST be rounded down before fulfilling the request.
+
+### Settlement Engine HTTP API
+
+#### Initiate an account
+
+Inform the settlement engine that a new accounting relationship was instantiated. The settlement engine MAY perform tasks as prerequisite to settle the new account. For example, a settlement engine implementation might send messages to the peer to exchange ledger identifiers or to negotiate terms of the settlement arrangement.
+
+The transport MUST be the only entity invoking this instruction to ensure the account identifier can be correlated with the correct peer to send and receive messages. Note that the transport also MUST have a mechanism to notify the accounting system that a new account should be instantiated (out-of-scope of this specification).
+
+##### Request
+
+```http
+POST /accounts/:id HTTP/1.1
+```
+
+##### Response
+
+```http
+HTTP/1.1 201 CREATED
+```
+
+#### Perform outgoing settlement
+
+Reconcile a liability owed to the peer for the given amount.
+
+Note that the settlement engine MAY accrue owed settlements without settling immediately due to varying arrangements, implementations, or conditions. The settlement engine SHOULD persist amounts owed to the peer that have yet to be settled so they can be settled later.
+
+The accounting system MUST be the only entity invoking this instruction and MUST preemptively debit this amount from the accounts payable, or some liability account tracking a balance owed to the peer, to ensure accurate accounting.
+
+##### Request
+
+```http
+POST /accounts/:id/settle HTTP/1.1
```

Comment:
In keeping with the RESTful design:

`POST /accounts/:id/settlements HTTP/1.1`

Reply:
I know ;)

Parts of this API don't make sense to me to model perfectly as a REST API. When a connector says to settle, there's no stateful "settlement" entity being created. When it says to send a message, there's no "state transfer": it's just a remote procedure call. IMO, we're using nouns to talk about things that are better described by verbs.

For example, the [LND REST API](https://api.lightning.community/rest/index.html) uses RESTful conventions where it makes sense to, and breaks from them where it doesn't.

Reply:
We could more strictly adhere to a RESTful design, but I think we sacrifice clarity. For example, the "message" endpoint would be the same for both the transport and SE, which seems confusing when they're performing different actions.

Reply:
[Kincaid regrettably gives in]

https://github.com/interledger/rfcs/compare/4e742ae54dcd738b67a4a541f9070837a3c16e98...ko-settlement-engine#diff-990a915317f4e35578edfa70631bd07dR265